# Semana 1

## Aula 01 - Conteúdo do Curso

O curso será apresentado por mim (Álvaro Justen) e Pedro Burgos e é dividido em
4 semanas:

- Contextualização, introdução a programação, instalação dos softwares
  necessários;
- Acesso a bases de dados, estatísticas básicas, criação de bases de dados;
- Aprofundamento na linguagem Python, estruturas de dados úteis, reutilização
  de código;
- Como a Web funciona, captura/extração de dados de site (Web scraping).

Esse é um curso introdutório, mostraremos os conceitos de maneira prática para
que você tenha um bom embasamento para continuar aprendendo depois dele. A cada
semana vamos apresentar um projeto aplicado ao jornalismo para fixar os
conceitos aprendidos.

### Dicas Para Acompanhar o Curso

- É muito importante ter acesso bom à Internet (sem bloqueios) e acesso
  administrativo ao computador (para instalar programas);
- Separe um tempo todos os dias para fazer as aulas - e só fazer isso durante
  esse tempo (nada de mídias sociais, ok?) -, isso lhe ajudará na memorização
  dos conceitos;
- Digite os comandos! Não copie/cole. Digitar é importante para e memorização;
- Interaja com os outros alunos! Utilize a plataforma e as mídias sociais para
  aprender conjuntamente (nossas contas no Twitter são:
  [@turicas](https://twitter.com/turicas) e
  [@burgos](https://twitter.com/burgos)).
- Marcaremos alguns encontros presenciais em Curitiba, São Paulo, Rio de
  Janeiro e talvez outras cidades! Acompanhe as novidades na plataforma para
  conhecer os outros alunos pessoalmente e trocar experiências. ;)

## Aula 02 - Contextualização e Instalação

### Contextualização

O *data pipeline* (ou funil de dados) é a sequência de passos que um jornalista
segue para fazer uma reportagem guiada por dados; são eles:

- Pergunte
- Encontre
- Obtenha
- Verifique
- Limpe
- Analise
- Apresente

Podemos utilizar a programação de computadores para auxiliar/automatizar os
processos a partir do "Obtenha", principalmente nos casos em que a quantidade
de dados é muito grande ou é necessária a repetição de muitas tarefas.  Saber
programar é saber dar ordens a um computador, que consegue facilmente executar
milhões de instruções a cada segundo.

A análise de dados não é algo novo: em 1890, [Herman
Hollerith](https://www.census.gov/history/www/census_then_now/notable_alumni/herman_hollerith.html)
criou uma máquina (mecânica, não era exatamente um computador) para processar
os dados do censo dos Estados Unidos - sem ela os estadunidenses demorariam
mais de dez anos (intervalo entre um censo e outro) para analisar esses dados!
Com ela, um operador conseguia processar milhares de cartões perfurados por
dia, que eram os "pendrives" da época.

Exemplo de matérias que utilizei programação para auxiliar na coleta e análise
de dados: [classificação de mais de 800 mil logradouros brasileiros por
gênero](http://www.generonumero.media/como-classificamo
-mais-de-800-mil-logradouros-brasileiros-por-genero/), pela [Revista Gênero e
Número](http://generonumero.media/).

Para instruirmos um computador precisamos falar uma linguagem de programação -
e existem centenas delas, cada uma com características específicas. As
linguagens de programação são como **ferramentas**: se você possui um prego,
utilizará um martelo (que é a ferramenta mais adequada para essa tarefa), porém
se possui um parafuso, utilizará uma chave Philips (que são melhores que as de
fenda ;).

A linguagem que aprenderemos nesse curso é [Python](http://www.python.org/) e
ela se destaca por ser fácil de aprender e utilizar, foi criada por um holandês
chamado [Guido Van Rossum](https://en.wikipedia.org/wiki/Guido_van_Rossum) no
começo da década de 90, é feita de maneira colaborativa (licenciada como
software livre). Python é largamente utilizada, possui uma comunidade de
desenvolvedores incrível que adora compartilhar código e conhecimento; existem
diversos eventos locais/regionais/nacionais em diversos países: procure a
comunidade/evento mais próximos e interaja com a galera!


### Instalação

Entre em [www.python.org](http://www.python.org/), clique em **Download** e
baixe a versão 3.6 para seu sistema operacional (está disponível para
GNU/Linux, Mac OS X e Windows) - procure pelo arquivo "executable installer".

> Nota (1): se você utiliza GNU/Linux é provável que o gerenciador de
> pacotes/loja de aplicativos de sua distribuição já tenha um pacote chamado
> "python3" - instale a partir dele, em vez de baixar no site.

> Nota (2): se você utiliza Windows, não esqueça de marcar a opção "Adicionar
> ao PATH" na primeira tela da instalação.

Após a instalação precisaremos testá-la para certificarmos de que você tem
todos os softwares necessários para o curso. Abra o interpretador de comandos do
seu sistema ("Terminal" no GNU/Linux e Mac OS X e "Prompt de Comando" no
Windows). Digite `python3` (GNU/Linux e MAC OS X) ou `python` (Windows) e
aperte *Enter*. Caso apareça uma mensagem "Python 3.6" e no fim os caracteres
">>>", significa que você instalou essa versão do Python com sucesso!  Agora,
basta digitar `exit()` e apertar *Enter* para sair do Python - mas ainda não
feche seu interpretador de comandos.

**Caso você tenha dificuldades na instalação, procure ajuda nos fóruns - é
imprescindível que essa etapa seja concluída com sucesso.**

Após a instalação do Python, instalaremos o Jupyter, que facilitará o processo
de desenvolvimento de programas Python. Ainda no seu interpretador de comandos
(mas já fora do Python), execute: `pip install jupyter` (digite esse comando
e aperte *Enter*).

> Nota: o comando acima baixará o Jupyter e todas suas dependências da Internet
> e poderá demorar um pouco para finalizar (depende de sua conexão).

Após tudo instalado, execute o comando: `jupyter notebook`. Ele irá mostrar um
endereço e o abrirá em seu navegador. A interface do Jupyter é acessada pelo
navegador: ele é um site, que roda no seu próprio computador. A partir de
agora, toda vez que você for fazer programas em Python você precisará do
Jupyter Notebook rodando - como tudo já foi instalado, basta abrir seu
interpretador de comandos e executar `jupyter notebook`.

> Nota: todos os comandos devem ser executados da forma como estão, pois letras
> maiúsculas e minículas são interpretadas de maneira diferente.

## Aula 03 - Lógica de Programação e Primeiros Programas

Os programas de computador são compostos por algoritmos, que são um conjunto de
instruções para o processador, são como o "modo de preparo" das receitas
culinárias. O funil de dados funciona como se fosse um algoritmo (bem genérico,
nesse caso): é uma série de passos para atingir um objetivo.

Imagine que uma instrução é uma frase que começa com um verbo e dá uma ordem ao
processador; ela precisa ser simples, passo-a-passo, não ambígua e
razoavelmente detalhada.  Exemplos de instruções do nosso dia-a-dia: inspire
por 5 segundos, feche os olhos.

Vamos a um exercício: escreva um algoritmo para descrever o processo de escovar
os dentes! Utilize o editor de textos que preferir (Microsoft Word, LibreOffice
Writer etc.) e escreva em Português frases simples, começadas por verbo
(instruções) com cada passo do processo de escovar os dentes.

Uma possível solução seria:

```
Se o tubo de pasta de dentes estiver vazio:
	Pegar um tubo novo na dispensa
Senão:
	Pegar o tubo de pasta de dentes no armário
Abrir o tubo
Pegar a escova de dentes
Colocar pasta na escova
Fechar o tubo
Abrir a torneira
Molhar a escova por 2 segundos
Fechar a torneira
Levar a escova à boca
Repetir para cada dente:
	Movimentar a escova de maneira circular
	Ir para o próximo dente
Escovar a língua
Abrir a torneira
Enquanto a boca estiver com pasta de dentes:
	Enxaguar a boca
Limpar a escova
Fechar a torneira
```

Pode ser que sua solução tenha ficado bastante diferente dessa, mas não tem
problema! Como a linguagem que utilizamos não possui muitas regras e
restrições, a possibilidade de existirem soluções completamente diferentes é
maior - em uma linguagem de programação isso não acontece nessa escala, dado
que as regras são mais restritas.

Repare em alguns detalhes:

- Utilizamos nomes de alguns objetos, como "tubo de pasta de dentes" e "escova
  de dentes". Na programação chamamos esses nomes de "variáveis" e elas
  referenciam uma determinada informação guardada na memória do computador;
- Ainda que você tivesse esquecido de especificar a instrução de fechar a
  torneira no final do algoritmo, ele cumpriria o papel de escovar os dentes,
  porém essa não seria uma boa prática ecologicamente falando, certo? Essa
  característica também acontece com os programas: programadores podem ser mais
  ou menos cuidadosos com o "ambiente" (memória do computador, tempo de
  processamento etc.);
- Talvez você tenha especificado as instruções em maior ou menor detalhe, por
  exemplo: utilizar várias instruções que juntas significam "abrir o tubo" ou
  utilizar apenas uma instrução para "abrir o tubo, pegar a escova de dentes,
  colocar pasta na escova, fechar o tubo". Nas linguagens de programação
  chamamos essa granularidade de "nível": uma linguagem é de nível "baixo"
  quando exige que você especifique cada detalhe e de nível "alto" quando você
  dá instruções mais genéricas e ela dá conta de traduzí-las em instruções mais
  detalhadas. Python é uma linguagem de altíssimo nível, que gera uma alta
  produtividade;
- Foram utilizadas estruturas de controle de fluxo: "e se..." e "enquanto...",
  que mudam a forma como o programa será executado, fazendo com que algumas
  instruções se repitam, sejam ou não executadas etc.

O processo de criação de um programa poderia ser resumir a:

- Entender o problema
- Sistematizar o algoritmo
- Escrever o código
- Executar/rodar o programa
- Verificar se o comportamento corresponde ao esperado


### Primeiro Programa em Python

Execute `jupyter notebook` em seu interpretador de comandos, espere ele abrir
no navegador e dentro dele clique em "New" e depois em "Python 3". A partir
desse momento, todos os comandos digitados no notebook serão comandos Python!
Vamos ao nosso primeiro programa:


In [1]:
print("Olá, Knight")

Olá, Knight


Após digitar o código basta apertar *Shift + Enter* e o resultado aparecerá
logo abaixo do código. Uma outra opção é apertar o botão *Run*. Em um Jupyter
Notebook você pode ter diversas "células" (caixas de texto) com códigos
diferentes. Caso queira alterar algum código basta clicar na célula, fazer a
alteração e executá-lo novamente (teclando *Shift + Enter* ou clicando em
*Run*).

A instrução `print` que utilizamos acima mostra um texto na tela; todos os
textos que especificarmos em Python deverão estar entre aspas.

> Curiosidade: `print` significa "imprimir". Essa instrução foi criada na época
> em que os computadores não possuíam monitores e a forma de saída padrão das
> informações era a impressora.


### Perguntando o Nome

Vamos perguntar o nome do usuário do programa e armazená-lo em uma variável
chamada "nome". A instrução `input` faz com que o Python pergunte alguma
informação ao usuário do programa, que deve responder digitando e apertando
*Enter*.

Digite na segunda célula e execute:


In [2]:
nome = input('Qual seu nome? ')
print('Olá, ' + nome)

Qual seu nome? Álvaro
Olá, Álvaro


No lado direito da primeira linha temos a instrução que pergunta o nome do
usuário. Essa instrução devolve o valor digitado e o sinal de "=" a armazena na
memória do computador, associando a ela o nome "nome". Na segunda linha
saudamos o usuário pelo nome e utilizamos o "+" para juntar os textos "Olá, " e
o que está contido na variável "nome".

Repare que "nome" foi um nome inventado por mim, não pelo Python. Poderíamos
escolher um outro, por exemplo:


In [3]:
nome_do_usuario = input('Qual seu nome? ')
print('Olá, ' + nome_do_usuario)

Qual seu nome? Álvaro
Olá, Álvaro


Toda variável deve começar com letras ou "\_" (*underscore*) e os outros caracteres podem ser
letras, "\_" e números, ou seja, espaços não são permitidos! Ah, atente-se às
maiúsculas e minúsculas: em Python, `nome_do_usuario` é diferente de
`Nome_do_Usuario`. Por convenção utilizamos nomes de variáveis minúsculas, com
as palavras separadas com "\_".

Além das instruções podemos adicionar comentários ao nosso código: são textos
quem explicam o funcionamento do código. Todos os comentários iniciam com
"#": Python irá ignorar tudo que vem depois desse símbolo até o fim da linha.
Vamos adicionar comentários ao código acima:



In [4]:
nome_do_usuario = input('Qual seu nome? ')  # Pede informação ao usuário
print('Olá, ' + nome_do_usuario)  # Mostra saudação com o nome

Qual seu nome? Álvaro
Olá, Álvaro


A partir de agora você pode utilizar os comentários em seus próximos programas!

## Aula 04 - Tipos de Dados

É muito importante entender a diferença entre os tipos de dados, afinal uma das
maiores aplicações da programação no jornalismo é justamente captura e análise
de dados.

> Nota: utilize o Jupyter Notebook em seu próprio computador e execute cada um
> dos exemplos mostrados aqui. Para facilitar a memorização dos comandos, **não
> copie e cole** código: digite cada letra, execute e, caso aconteça algum
> erro, verifique o que você digitou errado para aprender a maneira correta.

Vamos criar um programa simples para somar dois números:


In [5]:
valor1 = input('Digite um número: ')
valor2 = input('Digite outro número: ')
print(valor1 + valor2)

Digite um número: 42
Digite outro número: 10
4210


Ao executar o programa acima e responder às duas perguntas - digamos que eu
tenha respondido `42` e `10` - veremos um resultado inesperado na tela: `4210`;
o Python juntou os dois números, em vez de somá-los. O programa está correto,
porque não houve erro em sua execução, ou seja, escrevemos as instruções dentro
da sintaxe da linguagem, mas definitivamente não fez o que esperávamos! Esse é
um **erro de lógica**.

A instrução `input` acima nos devolve sempre um texto e o operador "+" junta
esses dois textos. Quer um exemplo? Responda "Oi, " na primeira pergunta e
"Python" na segunda: o resultado será "Oi, Python". O nome técnico para
variáveis que armazenam textos é "string" e no Python abreviamos e chamamos de
`str`. Em nossos programas, tudo que colocamos entre aspas (sejam elas simples
ou duplas) são strings.

O "+" pode significar a soma aritmética também, mas é necessário que as
variáveis que usarmos sejam números. Para que nosso programa funcione como
esperávamos devemos converter o que a instrução `input` nos devolve para
números inteiros - vamos utilizar a instrução `int` para isso:


In [6]:
valor1 = input('Digite um número: ')
valor2 = input('Digite outro número: ')
print(int(valor1) + int(valor2))

Digite um número: 42
Digite outro número: 10
52


Dessa forma, o resultado de `int(valor1)` será somado ao resultado de
`int(valor2)` e então o valor final será exibido na tela. Note que
nesse caso as variáveis `valor1` e `valor2` continuam sendo strings
e convertemos apenas para fazer a soma, na terceira linha do programa.
Alternativamente, poderíamos armazenar os valores já convertidos nas
variáveis, daí a conversão durante o `print` torna-se desnecessária:

In [7]:
valor1 = int(input('Digite um número: '))
valor2 = int(input('Digite outro número: '))
print(valor1 + valor2)

Digite um número: 42
Digite outro número: 10
52


Note que existem várias maneiras de resolver um mesmo problema.

Você lembra das funções durante as aulas de Matemática, como `g(x)`e
`f(g(x))`? Foi exatamente o que fizemos acima! A expressão `int(input('...'))`
é totalmente equivalente a `f(g(x))`. Na programação também chamamos de função
essas instruções que são seguidas de parenteses! A partir de agora as chamarei
de funções em vez de instruções, ok?

Recapitulando o fluxo do programa:

- A função `input` recebe um texto, mostra esse texto para o usuário, espera
  o usuário digitar algo e apertar *Enter*; após esse processo ela nos devolve
  uma string com o conteúdo que o usuário digitou;
- A função `int` recebe um texto e nos devolve o número inteiro correspondente
  àquele texto (exemplo: `int('42')` nos devolve `42`);
- No programa acima estamos pegando a resposta de `input`, passando para `int`
  e depois armazenando nas variáveis `valor1` e `valor2`.

Tente colocar um valor com casas decimais no programa acima e note que ele não
funcionará. O resultado será algo como:
`ValueError: invalid literal for int() with base 10: '3.14'`. Isso acontece
porque a função `int` só aceita textos que contém dígitos numéricos.

Além dos números inteiros Python suporta também números racionais,
representados pelo nome `float` (exemplo: `3.14`) e números complexos,
representados por `complex` (exemplo: `2 + 1.3j`). Note que utilizamos o ponto
e não a vírgula como separador de decimais.

Podemos, então, trocar `int` por `float` para que nosso programa suporte
números com casas decimais:

In [8]:
valor1 = float(input('Digite um número: '))
valor2 = float(input('Digite outro número: '))
print(valor1 + valor2)

Digite um número: 3.14
Digite outro número: 20
23.14


Com os tipos de dados numéricos conseguimos fazer as seguintes operações
artiméticas:

- Soma: `+`
- Subtração: `-`
- Multiplicação: `*`
- Divisão: `/`
- Exponenciação: `**`
- Resto da divisão inteira: `%`
- Divisão inteira: `//`

Além disso, a utilização de parenteses muda a precedência das operações,
exatamente como na Matemática! Exemplo: `(2 + 5) * 6` resulta em `42` (nesse
caso, primeiro é efetuada a soma e depois a multiplicação).

Agora vamos a um exercício: altere o programa acima para mostrar a média dos
dois números racionais, em vez da soma.

Resolução:

In [9]:
valor1 = float(input('Digite um número: '))
valor2 = float(input('Digite outro número: '))
print((valor1 + valor2) / 2)  # a média de 2 números é a soma dividida por 2

Digite um número: 20.5
Digite outro número: 10.5
15.5


Nós poderíamos melhorar um pouco nosso programa criando uma variável chamada
`media`, dessa forma ficaria claro que a expressão dentro do `print` é a média
de dois números:

In [10]:
valor1 = float(input('Digite um número: '))
valor2 = float(input('Digite outro número: '))
media = (valor1 + valor2) / 2
print(media)

Digite um número: 20.5
Digite outro número: 10.5
15.5


Ambos programas possuem o mesmo resultado, mas o segundo está mais legível.
Legibilidade do código é uma característica **muito importante** na
programação, pois agiliza o entendimento do código, fazendo com que possamos
modificar programas com mais rapidez e até evitando possíveis falhas.

Poderíamos alterar ainda mais o programa para que a mensagem exibida ao usuário
seja um pouco melhor:

In [11]:
valor1 = float(input('Digite um número: '))
valor2 = float(input('Digite outro número: '))
media = (valor1 + valor2) / 2
print('A média é: ' + str(media))

Digite um número: 20.5
Digite outro número: 10.5
A média é: 15.5


Repare que estamos convertendo `media` para string (usando a função `str`) e
então juntando a string resultante à string `'A média é: '` e com isso criamos
uma string contendo essa frase com o resultado do cálculo.

As estruturas de dados são parte importantíssima dos programas. Na próxima
semana aprenderemos outras dessas estruturas e como utilizá-las para manipular
bases de dados! Mas a semana ainda não acabou: na próxima aula aprenderemos
a tomar decisões em nossos programas.

## Aula 05 - Controle de Fluxo

Nessa aula vamos aprender uma das maneiras de fazer controle de fluxo em nossos
programas. O controle de fluxo dita a ordem em que as instruções serão
executadas, exemplos:

- Dada uma determinada condição, gostaríamos de executar certa instrução, porém
  caso essa condição não seja satisfeita, gostaríamos de executar outra
  instrução (como no algoritmo para escovar os dentes: se o tubo estiver vazio,
  busque na dispensa, senão busque no armário);
- Enquanto uma condição for satisfeita, execute determinada instrução (como no
  algoritmo para escovar os dentes: enquanto a escova não passar por cada
  dente, faça movimentos circulares e vá para o próximo dente).

Vamos a um exemplo de código: pediremos que o usuário digite um valor e,
dependendo do valor digitado iremos mostrar uma mensagem ou outra:



In [12]:
valor = float(input('Digite um número: '))
if valor > 9000:
    print('OVER 9K!')
else:
    print('Valor muito baixo =/')

Digite um número: 9000
Valor muito baixo =/


Tente rodar o programa mais de uma vez e digite números distintos!

Usamos o `if` para verificar o valor que o usuário digitou e para isso
utilizamos uma **expressão booleana** (`valor > 9000`), que é uma expressão que
devolverá um valor verdadeiro (`True`) ou falso (`False`).

Não esqueça de colocar as chamadas para a função `print` mais à direita. Esse
espaçamento para a direita chama-se **indentação** e serve para sinalizar ao
Python que essa instrução deve ser executada apenas se a condição dentro do
`if` for verdadeira; o mesmo acontece com o `else`, porém nesse caso a
instrução dentro dele será executada apenas se a condição do `if` for falsa.
O `else` é opcional.

Você pode colocar quantas instruções quiser dentro do `if` e dentro do `else`,
mas é necessário manter o mesmo nível de indentação. E caso queira seguir o
fluxo do programa (fora do `if` e  do `else`), basta voltar um nível de
indentação, exemplo:



In [16]:
valor = float(input('Digite um número: '))  # linha não indentada
if valor > 9000:  # linha não indentada
    print('OVER 9K!')  # linha indentada, dentro do if
    print('Esse texto também será exibido se valor > 9000.')  # indentada, dentro do if
else:  # linha não indentada
    print('Valor muito baixo =/')  # indentada, dentro do else
    print('Esse texto também será exibido se valor <= 9000.')  # indentada, dentro do else
valor2 = float(input('Digite um outro valor: '))  # não indentada

Digite um número: 9001
OVER 9K!
Esse texto também será exibido se valor > 9000.
Digite um outro valor: 42



No programa acima a linha `valor2 = ...` será executada independente de `valor`
ser maior que 9000.

Além do operador `>` podemos utilizar vários outros operadores que nos devolvem
`True` ou `False`, como:

- Menor que: `<`
- Maior que: `>`
- Menor ou igual a: `<=`
- Maior ou igual a: `>=`
- Igual: `==`
- Diferente: `!=`
- Ou: `or`
- E: `and`
- Negação: `not`

Repare no igual: utilizamos **dois símbolos** para verificar a igualdade, que é
diferente de quando queremos criar uma variável (o `=` é o operador de
atribuição a uma variável, já o `==` é o operador de verificação de igualdade).

`or`, `and` e `not` podem ser usados para agrupar expressões booleanas, dessa
forma conseguimos fazer diversas verificações em somente um `if`. Vamos
modificar o programa para imprimir "OVER 9K!" apenas se o usuário digitar dois
valores acima de 9000:

In [14]:
valor1 = float(input('Digite um número: '))
valor2 = float(input('Digite outro número: '))
if valor1 > 9000 and valor2 > 9000:
    print('OVER 9K!')
else:
    print('Valor muito baixo =/')

Digite um número: 9999
Digite outro número: 9000
Valor muito baixo =/


Agora temos duas expressões booleanas dentro do `if` e o `and` vai juntá-las,
devolvendo um novo valor (também `True` ou `False`). O padrão
`expressao1 and expressao2` devolverá `True` apenas se a `expressao1` for
`True` **e** a `expressao2` for `True`. Tente executar o programa acima com
os seguintes números e verifique o resultado:

- `9000` e `8999` (muito baixo)
- `9000` e `9999` (muito baixo)
- `9998` e `9000` (muito baixo)
- `9998` e `9999` (OVER 9K!)

Agora vamos a um exercício: modifique o programa da média para que ele se
recuse a utilizar números negativos (ou seja, se um dos números for menor que
zero).

Vamos à resolução:

In [15]:
valor1 = float(input('Digite um número: '))
valor2 = float(input('Digite outro número: '))
if valor1 < 0 or valor2 < 0:
    print('Me recuso!')
else:
    media = (valor1 + valor2) / 2
    print('A média é: ' + str(media))

Digite um número: -2
Digite outro número: 10
Me recuso!


Repare que dessa vez utilizamos o `or` em vez do `and`, pois queremos que o
programa se recuse caso `valor1` seja negativo **ou** caso `valor2` seja
negativo. Teste o programa com os seguintes números:

- `-2` e `-10` (deve se recusar)
- `-2` e `10` (deve se recusar)
- `2` e `-10` (deve se recusar)
- `2` e `10` (deve calcular)

Nessa aula aprendemos a tomar decisões em nossos programas com `if` e `else`!
Existem outras formas mais complexas de fazer controle de fluxo e as
aprenderemos ao longo das próximas semanas. :)